mono/metadata/marshal-ilgen.c

```diff
@@ -5182,35 +5199,61 @@ emit_marshal_safehandle_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 			init_safe_handle ();
 
 		if (t->byref){
-			ERROR_DECL (error);
-			MonoMethod *ctor;
-			
-			/*
-			 * My tests indicate that ref SafeHandles parameters are not actually
-			 * passed by ref, but instead a new Handle is created regardless of
-			 * whether a change happens in the unmanaged side.
-			 *
-			 * Also, the Handle is created before calling into unmanaged code,
-			 * but we do not support that mechanism (getting to the original
-			 * handle) and it makes no difference where we create this
-			 */
-			ctor = mono_class_get_method_from_name_checked (t->data.klass, ".ctor", 0, 0, error);
-			if (ctor == NULL || !is_ok (error)){
-				mono_mb_emit_exception (mb, "MissingMethodException", "paramterless constructor required");
-				mono_error_cleanup (error);
-				break;
+			/* If there was SafeHandle on input we have to release the reference to it */
+			if (IS_IN(t)) {
+				mono_mb_emit_ldloc (mb, dar_release_slot);
+				label_next = mono_mb_emit_branch (mb, CEE_BRFALSE);
+				mono_mb_emit_ldarg (mb, argnum);
+				mono_mb_emit_byte (mb, CEE_LDIND_I);
+				mono_mb_emit_managed_call (mb, sh_dangerous_release, NULL);
+				mono_mb_patch_branch (mb, label_next);
 			}
-			/* refval = new SafeHandleDerived ()*/
-			mono_mb_emit_ldarg (mb, argnum);
-			mono_mb_emit_op (mb, CEE_NEWOBJ, ctor);
-			mono_mb_emit_byte (mb, CEE_STIND_REF);
 
-			/* refval.handle = returned_handle */
-			mono_mb_emit_ldarg (mb, argnum);
-			mono_mb_emit_byte (mb, CEE_LDIND_REF);
-			mono_mb_emit_ldflda (mb, MONO_STRUCT_OFFSET (MonoSafeHandle, handle));
-			mono_mb_emit_ldloc (mb, conv_arg);
-			mono_mb_emit_byte (mb, CEE_STIND_I);
+			if (IS_OUT(t)) {
+				ERROR_DECL (error);
+				MonoMethod *ctor;
+			
+				/*
+				 * If the SafeHandle was marshalled on input we can skip the marshalling on
+				 * output if the handle value is identical.
+				 */
+				if (IS_IN(t)) {
+					int old_handle_value_slot = dar_release_slot + 1;
+					mono_mb_emit_ldloc (mb, old_handle_value_slot);
+					mono_mb_emit_ldloc (mb, conv_arg);
+					label_next = mono_mb_emit_branch (mb, CEE_BEQ);
+				}
+
+				/*
+				 * Create an empty SafeHandle (of correct derived type).
+				 * 
+				 * FIXME: If an out-of-memory situation or exception happens here we will
+				 * leak the handle. We should move the allocation of the SafeHandle to the
+				 * input marshalling code to prevent that.
+				 */
+				ctor = mono_class_get_method_from_name_checked (t->data.klass, ".ctor", 0, 0, error);
+				if (ctor == NULL || !is_ok (error)){
+					mono_mb_emit_exception (mb, "MissingMethodException", "paramterless constructor required");
+					mono_error_cleanup (error);
```

Comment:
use a `local_error`, otherwise this will poison `error` if it is set later (by the caller, for example).  Something like:
```c
    ERROR_DECL (local_error);
    ctor = get_method (..., local_error);
    if (ctor == NULL || !is_ok (local_error)) {
        ...
        mono_error_cleanup (local_error);
    }
```

Reply:
It is already local variable and the code is mostly unchanged from what was there before my change. I can rename it to `local_error` for additional clarity.

Reply:
Oh, sorry about that. Not sure what I looked at. This is fine, of course, if it's local